### PR TITLE
Test that all pipelines get the same data splits

### LINKED ISSUE #1982: Ensure all pipelines in AutoMLSearch receive the same data splits
Repro

```python
import joblib
from evalml.demos import load_fraud
from evalml.preprocessing.data_splitters import BalancedClassificationDataCVSplit

splitter = BalancedClassificationDataCVSplit(n_splits=3, random_seed=0, shuffle=True)

X, y = load_fraud(5000)
X = X.to_dataframe()
y = y.to_series().astype("int")

for train, test in splitter.split(X, y):
    print((joblib.hash(train), joblib.hash(test)))

# Output
('75f1b95d7ce307ac6c793055330969aa', '8c89fe1a592c50a700b6d5cbb02dba8b')
('f8c849bbfbed37c13f66c5c742e237cb', '9c4879fb550fded8be9ac03e95a1bf95')
('cdc21f0d6bbf45459c9695258f7f04dc', '5b575765bbe176e732b8eb4dc1bf2822')

for train, test in splitter.split(X, y):
    print((joblib.hash(train), joblib.hash(test)))

# Output
('bf462b82af243c552ac48acad2dfd748', '8c89fe1a592c50a700b6d5cbb02dba8b')
('b8341b536c63c7957c099b05e315f49c', '9c4879fb550fded8be9ac03e95a1bf95')
('780e74673b601790037fc0b17dde56fe', '5b575765bbe176e732b8eb4dc1bf2822')

for train, test in splitter.split(X, y):
    print((joblib.hash(train), joblib.hash(test)

# Output
('385f6c538568ad3a33cf84f61d94144c', '8c89fe1a592c50a700b6d5cbb02dba8b')
('8db65d0a3bdf87ae0f135b9766a260dd', '9c4879fb550fded8be9ac03e95a1bf95')
('2a7293fc1308b8a572091d7c76d20205', '5b575765bbe176e732b8eb4dc1bf2822')
```

This is different from the behavior of the sklearn splitter:

```python
from sklearn.model_selection import StratifiedKFold

kfold = StratifiedKFold(n_splits=3, random_state=0, shuffle=True)

for train, test in kfold.split(X, y):
    print((joblib.hash(train), joblib.hash(test)))

#Output
('6c30ee6a11803927024354405389506a', '8c89fe1a592c50a700b6d5cbb02dba8b')
('df0a70e2e6ca783f12461e8c82a26ad4', '9c4879fb550fded8be9ac03e95a1bf95')
('2898e4b3d3621b436641016499f4aafb', '5b575765bbe176e732b8eb4dc1bf2822')

for train, test in kfold.split(X, y):
    print((joblib.hash(train), joblib.hash(test)))

# Output
('6c30ee6a11803927024354405389506a', '8c89fe1a592c50a700b6d5cbb02dba8b')
('df0a70e2e6ca783f12461e8c82a26ad4', '9c4879fb550fded8be9ac03e95a1bf95')
('2898e4b3d3621b436641016499f4aafb', '5b575765bbe176e732b8eb4dc1bf2822')
```

I think this is problematic for two reasons:

1. Since BalancedClassificationDataCVSplit is the default splitter in automl, it means our pipelines are evaluated on different splits
2. Since `split` modifies the state of the data splitter it means we'll have different results between the sequential and parallel engines.

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1982

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/tests/automl_tests/test_automl.py [hunk 1]
```diff
@@ -4787,3 +4788,77 @@ def test_automl_thresholding_train_pipelines(mock_objective, threshold, X_y_bina
...
+    [
+        ProblemTypes.BINARY,
+        ProblemTypes.MULTICLASS,
+        ProblemTypes.REGRESSION,
+        ProblemTypes.TIME_SERIES_REGRESSION,
+        ProblemTypes.TIME_SERIES_BINARY,
+        ProblemTypes.TIME_SERIES_MULTICLASS,
+    ],
+)
+def test_data_splitter_gives_pipelines_same_data(
```
Comment (freddyaboulton) [inline evalml/tests/automl_tests/test_automl.py:4804]: Taken from the unit tests @dsherry added in #2210

File: evalml/tests/automl_tests/test_automl.py [hunk 2]
```diff
@@ -4787,3 +4788,77 @@ def test_automl_thresholding_train_pipelines(mock_objective, threshold, X_y_bina
...
+    n_splits = automl.data_splitter.n_splits
+    env = AutoMLTestEnv(automl_type)
+    with env.test_context(score_return_value={automl.objective.name: 1.0}):
+        automl.search()
+    n_pipelines_evaluated = len(automl.results["pipeline_results"])
+    assert n_pipelines_evaluated > 1
+
+    # current automl algo trains each pipeline using 3-fold CV for "small" datasets (i.e. test data above)
+    # therefore, each pipeline should recieve an identical set of three training-validation splits
+    X_fit_hashes, y_fit_hashes = defaultdict(set), defaultdict(set)
```
Comment (bchen1116) [inline evalml/tests/automl_tests/test_automl.py:4838]: This is super cool!

**Review by bchen1116 — APPROVED** [review]
Looks good! Nice test coverage!

**Review by chukarsten — APPROVED** [review]
That was a super big brained move with the hashes and the assertion.  It took me longer than I care to admit to understand how this worked, lol.

**Review by angela97lin — APPROVED** [review]
Love this addition, LGTM 👍

**Pull request merged**